include/private/hashmap.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Hash map support
+ *
+ * Copyright 2020 Philip Rebohle for Valve Corporation
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#ifndef __VKD3D_HASHMAP_H
+#define __VKD3D_HASHMAP_H
+
+#include <stddef.h>
+
+#include "memory.h"
+
+enum hash_map_entry_flag
+{
+    HASH_MAP_ENTRY_OCCUPIED = (1 << 0),
+};
+
+struct hash_map_entry
+{
+    uint32_t hash_value;
+    uint32_t flags;
+};
+
+typedef uint32_t (*pfn_hash_func)(const void* key);
+typedef bool (*pfn_hash_compare_func)(const void *key, const struct hash_map_entry *entry);
+
+/* Open-addressing hash table */
+struct hash_map
+{
+    pfn_hash_func hash_func;
+    pfn_hash_compare_func compare_func;
+    void *entries;
+    size_t entry_size;
+    uint32_t entry_count;
+    uint32_t used_count;
+};
+
+static inline struct hash_map_entry *hash_map_get_entry(struct hash_map *hash_map, uint32_t entry_idx)
+{
+    return hash_map->entries + hash_map->entry_size * entry_idx;
```

Comment:
This breaks MSVC build because it disallows pointer arithmetic on `void *`. The exact error is:
```
       "Z:\home\saancreed\Repositories\vkd3d\build\vkd3d.sln" (default target) 
       (1) ->
       "Z:\home\saancreed\Repositories\vkd3d\build\libs\vkd3d\eba9057@@vkd3d@st
       a.vcxproj" (default target) (7) ->
       (ClCompile target) -> 
         Z:\home\saancreed\Repositories\vkd3d\include\private\hashmap.h(55,64):
        error C2036: 'void *': unknown size [Z:\home\saancreed\Repositories\vkd
       3d\build\libs\vkd3d\eba9057@@vkd3d@sta.vcxproj]
```

Reply:
Yes, pointer arithmetic on `void *` is a GNU extension.

Reply:
Fixed.